src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -277,7 +270,7 @@ private void Release(int releaseCount)
                         {
                             Interlocked.Increment(ref _successfulLeasesCount);
                         }
-                        nextPendingRequest.CancellationTokenRegistration.Dispose();
+                        disposer.Add(nextPendingRequest);
```

Comment:
Was just double checking things and realized the `Debug.Assert(_queueCount == 0);` on line 285 can now be wrong (for other limiters too).

I'm kind of tempted to just remove it. Maybe add some asserts elsewhere?

Reply:
And other asserts about `_queueCount` too. I don't know what dotnet/runtime policy on asserts is, i.e. how important it is to keep them. It should be possible to keep these asserts as is by moving `_queueCount -= registration.Count` inside the lock in case of disposal, and setting `registration.Count` to zero (it's only accessed from within the lock so it's safe to modify), but it makes the code a bit more complicated.

Reply:
I don't think this one is extremely important. It's there to verify that when the limiter is idle, there are no queued items.

We could check that the `_queue` is empty instead, we do lose the verification that our math is correct, but like you noted, it would be more complex to verify that.

Reply:
Not as complex as I thought. How about something like this?

```diff
diff --git a/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs b/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs
index c2ba9cac..4faf4210 100644
--- a/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs
+++ b/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs
@@ -178,7 +178,7 @@ protected override ValueTask<RateLimitLease> AcquireAsyncCore(int permitCount, C
                 var request = new RequestRegistration(permitCount, this, cancellationToken);
                 _queue.EnqueueTail(request);
                 _queueCount += permitCount;
-                Debug.Assert(_queueCount <= _options.QueueLimit);
+                Debug.Assert(_queueCount - disposer.QueueCount <= _options.QueueLimit);
 
                 return new ValueTask<RateLimitLease>(request.Task);
             }
@@ -271,7 +271,7 @@ private void Release(int releaseCount)
                             Interlocked.Increment(ref _successfulLeasesCount);
                         }
                         disposer.Add(nextPendingRequest);
-                        Debug.Assert(_queueCount >= 0);
+                        Debug.Assert(_queueCount - disposer.QueueCount >= 0);
                     }
                     else
                     {
@@ -282,7 +282,7 @@ private void Release(int releaseCount)
                 if (_permitCount == _options.PermitLimit)
                 {
                     Debug.Assert(_idleSince is null);
-                    Debug.Assert(_queueCount == 0);
+                    Debug.Assert(_queueCount - disposer.QueueCount == 0);
                     _idleSince = Stopwatch.GetTimestamp();
                 }
             }
@@ -424,10 +424,13 @@ public struct Disposer : IDisposable
             {
                 private RequestRegistration? _next;
 
+                public int QueueCount { get; set; }
+
                 public void Add(RequestRegistration request)
                 {
                     request._next = _next;
                     _next = request;
+                    QueueCount += request.Count;
                 }
 
                 public void Dispose()
```

Reply:
Oh that's not bad! Maybe a comment on `public int QueueCount { get; set; }` that it's just for asserts

Reply:
I guess this doesn't work because the cancellation callback might have already run, and the `_queueCount` will have already been decreased. This now can double count.

Reply:
No, it didn't. I'll make a different change. Was a bit busy.

Reply:
Outstanding leases' `Count` properties are only accessed under rate limiter lock, so this double counting is easily prevented.

Reply:
> Process terminated. Assertion failed.
   at System.Threading.RateLimiting.TokenBucketRateLimiter.ReplenishInternal(Int64 nowTicks) in /_/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs:line 358

Reply:
Sorry for wasting your itme. I got confused by your initial comment about the `_queueCount` assert. Actually disposing the `RegistrationRequest` has no effect on `_queueCount`, as it only disposes the `CancellationTokenRegistration` and does not run `RegistrationRequest.Cancel()`, so the call to `RegistrationRequest.Dispose()` can be moved with no issues. I've reverted the last two changes. All RateLimiter unit tests pass now.

Reply:
Completely reworked `_queueCount` handling and added unit tests exercising the cancellation code paths which used to produce the deadlock (only in ConcurrentLimiter, other three implementations are identical in this respect).

Reply:
This is getting a bit complicated. Can we just remove `Debug.Assert(_queueCount == 0);` on line 285 and undo the recent changes?

Reply:
@atykhyy Are you able to update the PR or should I try to finish it up? We have a week or two before 8.0 starts locking down and I want to get this change in 😄

Reply:
Sorry, I was out of office. I don't quite understand why you wanted to remove this assert. It's not hit in unit tests, and I think it can be proved that it is safe. But if you're okay removing it then I won't bother :)